### PR TITLE
Calculate csm on low-res image

### PR DESCRIPTION
**Pull request by ckolbPTB**

### CODE REVIEW

File: src/mrpro/data/CsmData.py [hunk 1]
```diff
@@ -44,6 +75,10 @@ def from_kdata_walsh(
         chunk_size_otherdim
             How many elements of the other dimensions should be processed at once.
             Default is `None`, which means that all elements are processed at once.
+        downsampled_size
+            IData will be downsampled to this size before calculating the csm to speed up the calculation and
+            reduce memory requirements. The final csm will be upsampled to the original size. If set to None no
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
            reduce memory requirements. The final csm will be upsampled to the original size. If set to `None` no
```

File: src/mrpro/data/CsmData.py [hunk 2]
```diff
@@ -75,19 +114,24 @@ def from_idata_walsh(
         chunk_size_otherdim:
             How many elements of the other dimensions should be processed at once.
             Default is `None`, which means that all elements are processed at once.
+        downsampled_size
+            IData will be downsampled to this size before calculating the csm to speed up the calculation and
+            reduce memory requirements. The final csm will be upsampled to the original size. If set to None no
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
            reduce memory requirements. The final csm will be upsampled to the original size. If set to `None` no
```

File: src/mrpro/data/CsmData.py [hunk 3]
```diff
@@ -117,6 +162,10 @@ def from_kdata_inati(
         chunk_size_otherdim
             How many elements of the other dimensions should be processed at once.
             Default is `None`, which means that all elements are processed at once.
+        downsampled_size
+            IData will be downsampled to this size before calculating the csm to speed up the calculation and
+            reduce memory requirements. The final csm will be upsampled to the original size. If set to None no
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
            reduce memory requirements. The final csm will be upsampled to the original size. If set to `None` no
```

File: src/mrpro/data/CsmData.py [hunk 4]
```diff
@@ -148,15 +201,26 @@ def from_idata_inati(
         chunk_size_otherdim:
             How many elements of the other dimensions should be processed at once.
             Default is None, which means that all elements are processed at once.
+        downsampled_size
+            IData will be downsampled to this size before calculating the csm to speed up the calculation and
+            reduce memory requirements. The final csm will be upsampled to the original size. If set to None no
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
            reduce memory requirements. The final csm will be upsampled to the original size. If set to `None` no
```

File: src/mrpro/data/CsmData.py [hunk 5]
```diff
@@ -148,15 +201,26 @@ def from_idata_inati(
         chunk_size_otherdim:
             How many elements of the other dimensions should be processed at once.
             Default is None, which means that all elements are processed at once.
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
            Default is `None`, which means that all elements are processed at once.
```

File: src/mrpro/utils/__init__.py [hunk 6]
```diff
@@ -12,14 +12,18 @@
 from mrpro.utils.sliding_window import sliding_window
 from mrpro.utils.reshape import broadcast_right, unsqueeze_left, unsqueeze_right, reduce_view, reshape_broadcasted, ravel_multi_index, unsqueeze_tensors_left, unsqueeze_tensors_right, unsqueeze_at, unsqueeze_tensors_at
 from mrpro.utils.TensorAttributeMixin import TensorAttributeMixin
+from mrpro.utils.zero_pad_or_crop import zero_pad_or_crop
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply src/mrpro/utils/__init__.py]: fixed

File: src/mrpro/utils/interpolate.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    if len(dim) != len(size):
+        raise ValueError('Must provide matching length size and dim arguments.')
+
+    # normalize dim to allow negative indexing in input
+    dim = tuple([a % x.ndim for a in dim])
+    if len(dim) != len(set(dim)):
+        raise ValueError(f'Dim must be unique. Normalized dims are {dim}')
+
+    # return input tensor if old and new size match
+    if all(x.shape[d] == s for s, d in zip(size, dim, strict=True)):
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply src/mrpro/utils/interpolate.py:39]: Ok, python version is >3.9 in the repository.

File: tests/utils/test_interpolate.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    return data
+
+
+@pytest.mark.parametrize('size', [10, 20, 30])
+@pytest.mark.parametrize('data_dtype', [torch.float32, torch.float64, torch.complex64, torch.complex128])
+def test_interpolate_linear(data: torch.Tensor, size: int, data_dtype: torch.dtype) -> None:
+    """Linear ramp should remain a linear ramp after inear interpolation."""
+    result = interpolate(data.to(dtype=data_dtype), size=(size,), dim=(4,), mode='linear')
+    assert result.dtype == data_dtype
+    assert torch.sum((torch.diff(result[..., 1:-1], dim=-1) - data.shape[-1] / size).abs()) < 1e-2
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply tests/utils/test_interpolate.py]: torch.diff(result[..., 1:-1], dim=-1).isclose(torch.tensor(data_shape[-1] / size)).all() ?

File: tests/utils/test_interpolate.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+def test_apply_lowres(data: torch.Tensor) -> None:
+    """Applying identity function should not change linear ramp data except for edges."""
+    data = data.to(torch.complex64)
+
+    def identity_function(x: torch.Tensor):
+        return x
+
+    apply = apply_lowres(identity_function, size=(8, 8, 8), dim=(-3, -2, -1))
+    torch.testing.assert_close(data[..., 1:-1], apply(data)[..., 1:-1])
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply tests/utils/test_interpolate.py:52]: matches docstring, expecter behavior.

File: tests/data/test_csm_data.py [hunk 10]
```diff
@@ -51,6 +55,24 @@ def test_CsmData_cuda(csm_method, ellipse_phantom, random_kheader) -> None:
...
+    csm_method: Callable, ellipse_phantom: EllipsePhantomTestData, random_kheader: KHeader
+) -> None:
+    """CsmData downsampling does not change smooth coil sensitivity maps."""
+    idata, csm_ref = multi_coil_image(n_coils=4, ph_ellipse=ellipse_phantom, random_kheader=random_kheader)
+
+    # Estimate coil sensitivity maps on original image size
+    smoothing_width = SpatialDimension(z=1, y=3, x=3)
+    csm = csm_method(idata, smoothing_width)
+
+    # Estimate coil sensitivity maps on low resolution image
```
Comment (fzimmermann89) [inline tests/data/test_csm_data.py]:
```suggestion
```
The comments just directly describe what the next line does anyways, right?

Instead, maybe comment why smoothing_width set to 1?
And name csm to csm_from_original `?

File: src/mrpro/data/CsmData.py [hunk 11]
```diff
@@ -11,13 +11,43 @@
...
+    -------
+        Size of downsampled data
+    """
+    if downsampled_size is None:
+        return (data_size[-3], data_size[-2], data_size[-1])  # needed for mypy
+
+    if isinstance(downsampled_size, int):
+        downsampled_size = SpatialDimension(z=downsampled_size, y=downsampled_size, x=downsampled_size)
+
+    downsampled_size.z = min(downsampled_size.z, data_size[-3])
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply src/mrpro/data/CsmData.py]: I think this could become an issue, yes

File: src/mrpro/data/CsmData.py [hunk 12]
```diff
@@ -11,13 +11,43 @@
...
+    downsampled_size.y = min(downsampled_size.y, data_size[-2])
+    downsampled_size.x = min(downsampled_size.x, data_size[-1])
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
    return (
        min(downsampled_size.z, data_size[-3]),
        min(downsampled_size.y, data_size[-2]),
        min(downsampled_size.x, data_size[-1]),
    )
```

File: src/mrpro/data/CsmData.py [hunk 13]
```diff
@@ -11,13 +11,43 @@
...
+
+    return downsampled_size.zyx
```
Comment (fzimmermann89) [inline src/mrpro/data/CsmData.py]:
```suggestion
```

**Review by fzimmermann89 — APPROVED** [review]
minor changes

File: tests/data/test_csm_data.py [hunk 14]
```diff
@@ -51,6 +55,24 @@ def test_CsmData_cuda(csm_method, ellipse_phantom, random_kheader) -> None:
...
+    csm_from_lowres_image = csm_method(idata, downsampled_size=idata.data.shape[-1] // 6, smoothing_width=1)
```
[the remark this answers is not in the document]
Comment (fzimmermann89) [reply tests/data/test_csm_data.py]: is there a reason for it?

[the remark this answers is not in the document]
Comment (ckolbPTB) [reply tests/data/test_csm_data.py]: The idea was to compare "Estimate coil sensitivity maps on original image size with smoothing width = 3" to "Estimate coil sensitivity maps on 3x downsampled image with smoothing width = 1"
